File: python/examples/hello_pub.py
#!/usr/bin/env python3
"""Publish HelloWorld samples on domain 0.

Usage:
    python examples/hello_pub.py [--domain DOMAIN_ID] [--count COUNT]

rtiddsspy should display the published samples.
"""

import sys
import os
import time
import argparse
import logging

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from vibedds.participant import DomainParticipant
from vibedds.qos import QosPolicy, ReliabilityKind
from vibedds.type_support import HelloWorldType


def main():
    parser = argparse.ArgumentParser(description="HelloWorld Publisher")
    parser.add_argument("--domain", type=int, default=0)
    parser.add_argument("--count", type=int, default=0, help="0 = infinite")
    parser.add_argument("--interval", type=float, default=1.0)
    parser.add_argument("-v", "--verbose", action="store_true")
    args = parser.parse_args()

    logging.basicConfig(
        level=logging.DEBUG if args.verbose else logging.INFO,
        format="%(asctime)s %(name)s %(levelname)s: %(message)s",
    )

    dp = DomainParticipant(domain_id=args.domain)
    dp.start()

    topic = dp.create_topic("HelloWorldTopic", HelloWorldType.TYPE_NAME)
    writer = dp.create_writer(topic, QosPolicy(reliability=ReliabilityKind.BEST_EFFORT))

    print(f"HelloWorld Publisher started on domain {args.domain}")
    print(f"  GUID: {dp.guid_prefix}")
    print(f"  Topic: {topic.name}")
    print("Press Ctrl+C to stop.\n")

    # Send initial SPDP announcement
    dp.announce_spdp()

    try:
        n = 0
        last_write = 0.0
        while True:
            dp.spin_once(timeout=0.05)

            now = time.time()
            if now - last_write >= args.interval:
                n += 1
                message = f"Hello DDS World #{n}"
                payload = HelloWorldType.serialize(message)
                writer.write(payload)
                print(f"Published: {message}")
                last_write = now

                if args.count > 0 and n >= args.count:
                    break
    except KeyboardInterrupt:
        print("\nStopping...")
    finally:
        dp.stop()


if __name__ == "__main__":
    main()

File: rust/vibedds/examples/hello_pub.rs
/// HelloWorld publisher example - publishes HelloWorld messages
/// that can be received by rtiddsspy or hello_sub.
///
/// Run: cargo run --example hello_pub
/// Then in another terminal: rtiddsspy -domainId 0 -printSample
/// Or: cargo run --example hello_sub

use std::time::Duration;

use vibedds::participant::DomainParticipant;
use vibedds::qos::QosPolicy;
use vibedds::type_support::HelloWorldType;

fn main() {
    // Initialize logging
    env_logger::Builder::from_env(env_logger::Env::default().default_filter_or("info"))
        .format_timestamp_millis()
        .init();

    println!("VibeDDS HelloWorld Publisher");
    println!("=============================");
    println!("Publishing HelloWorld messages.");
    println!("Run rtiddsspy -printSample to see the data.");
    println!();

    // Create participant
    let mut participant = DomainParticipant::new(0, 0);
    participant.set_spdp_announce_interval(Duration::from_secs(5));

    // Start the participant
    participant.start().expect("Failed to start participant");
    println!("Participant started: {:?}", participant.guid_prefix());

    // Send initial SPDP announcement
    participant.announce_spdp().expect("Failed to announce SPDP");

    // Create a HelloWorld DataWriter
    let writer_qos = QosPolicy::reliable();
    let writer_arc = participant.create_writer(
        "HelloWorld",
        "HelloWorld",
        writer_qos,
    );
    println!("Created HelloWorld writer");

    // Set up discovery callback
    participant.on_participant_discovered(|p| {
        println!(
            "Discovered participant: {:?}, vendor={:?}",
            p.guid_prefix, p.vendor_id
        );
    });

    println!();
    println!("Publishing every 2 seconds... Press Ctrl+C to stop.");
    println!();

    let mut count = 0u32;
    let mut last_publish = std::time::Instant::now();

    // Run the event loop
    loop {
        if let Err(e) = participant.spin_once(Duration::from_millis(100)) {
            log::error!("Error in spin: {}", e);
        }

        // Periodically announce SEDP
        if let Err(e) = participant.announce_sedp() {
            log::error!("Error in SEDP announce: {}", e);
        }

        // Publish every 2 seconds
        if last_publish.elapsed() >= Duration::from_secs(2) {
            count += 1;
            let message = format!("Hello DDS World #{}", count);
            let payload = HelloWorldType::serialize(&message);

            {
                let mut writer = writer_arc.lock().unwrap();
                if let Err(e) = participant.send_data(&mut writer, &payload) {
                    log::error!("Failed to send data: {}", e);
                } else {
                    println!("Published: {}", message);
                }
            }

            last_publish = std::time::Instant::now();
        }

        std::thread::sleep(Duration::from_millis(50));
    }
}
